### PR TITLE
Add distributed cache

### PR DESCRIPTION
**Pull request by bunyodkamalov**

### CODE REVIEW

File: src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/EntityRepositoryBase.cs [hunk 1]
```diff
@@ -1,19 +1,35 @@
...
 {
-    private TContext DbContext => _dbContext;
-    private readonly TContext _dbContext;
-
-    protected EntityRepositoryBase(TContext dbContext)
-    {
-        _dbContext = dbContext;
-    }
-
+    private TContext DbContext => dbContext;
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/EntityRepositoryBase.cs:25]: Make this protected

File: src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/EntityRepositoryBase.cs [hunk 2]
```diff
@@ -25,95 +41,197 @@ protected IQueryable<TEntity> Get(Expression<Func<TEntity, bool>>? predicate = d
...
+    }
+
+    /// <summary>
+    ///  Asynchronously retrieves entities from the repository based on a query specification and caching options.
+    /// </summary>
+    /// <param name="querySpecification"></param>
+    /// <param name="asNoTracking"></param>
+    /// <param name="cancellationToken"></param>
+    /// <returns></returns>
+    protected async ValueTask<IList<TEntity>> GetAsync(QuerySpecification<TEntity> querySpecification, bool asNoTracking = false,
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/EntityRepositoryBase.cs:53]: Add as no tracking to query specification and remove from method parameters

File: src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/EntityRepositoryBase.cs [hunk 3]
```diff
@@ -25,95 +41,197 @@ protected IQueryable<TEntity> Get(Expression<Func<TEntity, bool>>? predicate = d
...
+
+            foundEntities = await initialQuery.ToListAsync(cancellationToken);
+
+            if (cacheEntryOptions is not null)
+                await cacheBroker.SetAsync(cacheKey, foundEntities, cacheEntryOptions);
+        }
+        else
+        {
+            foundEntities = cachedEntity;
+        }
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/EntityRepositoryBase.cs:77]: Don't need brackets here

File: src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/EntityRepositoryBase.cs [hunk 4]
```diff
@@ -25,95 +41,197 @@ protected IQueryable<TEntity> Get(Expression<Func<TEntity, bool>>? predicate = d
...
+
+            foundEntity = await initialQuery.FirstOrDefaultAsync(entity => entity.Id == id, cancellationToken);
+
+            if (cacheEntryOptions is not null && foundEntity is not null)
+                await cacheBroker.SetAsync(foundEntity.Id.ToString(), foundEntity, cacheEntryOptions);
+        }
+        else
+        {
+            foundEntity = cachedEntity;
+        }
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/EntityRepositoryBase.cs:112]: No need for brackets here

File: src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/Interfaces/IUserRepository.cs [hunk 5]
```diff
@@ -16,6 +17,16 @@ public interface IUserRepository
...
 
+    /// <summary>
+    /// /// Retrieves a collection of users based on the specified predicate. 
+    /// </summary>
+    /// <param name="querySpecification"></param>
+    /// <param name="asNoTracking"></param>
+    /// <param name="cancellationToken"></param>
+    /// <returns></returns>
+    ValueTask<IList<User>> GetAsync(QuerySpecification<User> querySpecification, bool asNoTracking = false,
+        CancellationToken cancellationToken = default);
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/Interfaces/IUserRepository.cs:28]: Remove as no tracking

File: src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Identity/Services/IUserService.cs [hunk 6]
```diff
@@ -25,6 +26,16 @@ public interface IUserService
...
 
+    /// <summary>
+    /// Retrieves a collection of users based on the specified predicate.
+    /// </summary>
+    /// <param name="querySpecification"></param>
+    /// <param name="asNoTracking"></param>
+    /// <param name="cancellationToken"></param>
+    /// <returns></returns>
+    ValueTask<IList<User>> GetAsync(QuerySpecification<User> querySpecification, bool asNoTracking = false,
+        CancellationToken cancellationToken = default);
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Identity/Services/IUserService.cs:37]: Remove as no tracking

File: src/Training.AirBnb.Clone.Backend/AirBnB.Api/Controllers/AccountsController.cs [hunk 7]
```diff
@@ -11,14 +12,16 @@ namespace AirBnB.Api.Controllers;
...
     [HttpGet]
-    public async ValueTask<IActionResult> Get()
+    public async ValueTask<IActionResult> Get([FromQuery] FilterPagination filterPagination)
     {
-        var users = await userService.Get(asNoTracking: true).ToListAsync();
-        var result = mapper.Map<IEnumerable<UserDto>>(users);
-        return result.Any() ? Ok(result) : NoContent();
-    }
+        var specification = new QuerySpecification<User>(filterPagination.PageSize, filterPagination.PageToken, true);
+        var result = await userService.GetAsync(specification);
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Api/Controllers/AccountsController.cs:20]: Don't create query specification manually, use `IQueryConvertible` interface method

**Review by sulton-max — CHANGES_REQUESTED** [review]
Good job

**Review by botirjonabdumannonovv — APPROVED** [review]

**Review by botirjonabdumannonovv — DISMISSED** [review]

**Review by sulton-max — COMMENTED** [review]
Good just change filter to query inside action

**Review by botirjonabdumannonovv — APPROVED** [review]

**Review by sulton-max — APPROVED** [review]

**Pull request merged**